src/core/egglog-herbie.rkt

```diff
@@ -854,54 +847,106 @@
-       (if (or (equal? last-line "true") (> total_nodes node-limit))
-           (sub1 curr-iter)
-           (loop (add1 curr-iter)))])))
+       ;; Run the ruleset once more
+       (define math-schedule
+         (list '(push)
+               `(run-schedule (repeat 1 ,tag))
+               '(print-size)
+               '(run unsound-rule 1)
+               '(extract (unsound))))
+
+       ;; Get egglog output
+       (define-values (math-exit? math-nodes)
+         (get-egglog-output math-schedule egglog-process egglog-output egglog-in err node-limit))
+
+       (cond
+         ;;  Check for saturation we verify the number of nodes stayed the same from a previous iteration
+         ;;  If so, return the iter-limit and do not need to run again
+         [(equal? math-nodes prev-number-nodes) (values curr-iter #f)]
+
+         ;; If Unsoundness detected or node-limit reached, then return the
+         ;; optimal iter limit (one less than current)
+         [math-exit?
+          ;; Pop once at the end since the egraph isn't valid
+          (send-to-egglog (list '(pop)) egglog-process egglog-output egglog-in err)
+
+          ;; Return one less than current iteration and indicate that we need to run again because pop
+          (values (sub1 curr-iter) #t)]
+
+         ;; Continue to next iteration of the math rules
+         [else
+          ;; set for the constant folding iteration
+          (set! prev-number-nodes math-nodes)
+
+          ;; 3. Run const-fold
+          (define const-schedule
+            (list '(push)
+                  `(run-schedule (repeat 1 const-fold))
+                  '(print-size)
+                  '(run unsound-rule 1)
+                  '(extract (unsound))))
+
+          (define-values (const-exit? const-nodes)
+            (get-egglog-output const-schedule egglog-process egglog-output egglog-in err node-limit))
+
+          (cond
+            [(equal? const-nodes prev-number-nodes) (values curr-iter #f)]
+            [const-exit?
+             (send-to-egglog (list '(pop)) egglog-process egglog-output egglog-in err)
+             (values (sub1 curr-iter) #t)]
+
+            [else
+             ;; Update state for the next iteration
+             (set! prev-number-nodes const-nodes)
+
+             (loop (add1 curr-iter))])])])))
+
+(define (get-egglog-output curr-schedule egglog-process egglog-output egglog-in err node-limit)
+  (define-values (node-values unsound?)
+    (send-to-egglog-unsound-detection curr-schedule egglog-process egglog-output egglog-in err))
+
+  ;  (when unsound?
+  ;    (printf "ALERT : UNSOUNDNESS DETECTED when...\n"))
+
+  (define total_nodes (calculate-nodes node-values))
+
+  ;; There are 3 cases when we can exit the unsound detection
+  ;;  1. Unsoundness is detected in the egraph
+  ;;  2. We have reached or exceeded the set node limit
+  ;;  3. Saturation check which is done in parent function
+  (define exit-detect? (or unsound? (>= total_nodes node-limit)))
+
```

Comment:
I'm realizing that checking nodes isn't quite the right thing to do. For example, you could theoretically have a ruleset that merges e-classes but doesn't add any nodes. That ruleset has made progress but this logic considers it to be saturated. Might be a feature request to the egglog team to not have to hack this in. I say this is acceptable for now but would appreciate a comment in the code!

Reply:
That also isn't precise enough because there are e-classes that may be introduced that don't have names